certbot-ci/certbot_integration_tests/utils/acme_server.py

```diff
@@ -20,27 +20,31 @@
 HTTP_01_PORT = 5002
 
 
-def setup_acme_server(acme_server, nodes):
+def setup_acme_server(acme_server, nodes, proxy=True):
```

Comment:
Yes it would continue to work. But it is very limited, since the teardown occurs only when the main python process finishes. I have some use cases in mind that could require the teardown to occur before. 

Here for now, the teardown is connected to the pytest general teardown, so calling programmatically pytest (as described in the official documentation https://docs.pytest.org/en/latest/usage.html#calling-pytest-from-python-code) will work as expected, with a teardown occurring at the end of the pytest invocation, not at the end of the python script.

Reply:
> Yes it would continue to work. But it is very limited, since the teardown occurs only when the main python process finishes. I have some use cases in mind that could require the teardown to occur before.

What's the use case you had in mind? If it requires further changes to our code, what do you think about waiting to make this change until that use case comes up? I just think sticking with `atexit` is cleaner and simpler given the current needs I'm aware of.

Reply:
This https://github.com/certbot/certbot/pull/7073#discussion_r289206742 for instance ^^

Reply:
Right, but you're pushing back on making that change and like I said, I'd rather not add options unless they're really needed.

Any other real, current use cases you have in mind?

Reply:
OK, last argument on my side. If you do not agree after it, I revert to atexit.

It is true that there is no actual use in certbot codebase, server teardown will work correctly with atexit. Two things pushed me to make this change. First one is, as I said before, potential uses of this logic in other contexts in certbot.

Second one is that the reason we did this PR, so to offer a way to have an alternative to the bash scripts that will be removed, made me realized that Certbot integration tests could be run by other projects. I would expect for instance that other ACME server projects are using some tests against certbot, and maybe reuse parts of our own tests. 

So in the scope of external usages, I do not like a lot the atexit approach, as we do not offer a way to control when the teardown occurs: it is at the end of the python process, independantly from what is happening during the script itself. I already talked about pytest programmatical call on that matter.

Here I give a way to control this teardown. Future useful API could be also to provide a context manager, even if with these modifications, it is really easy to do it by yourself. 

And finally, the modifications done to implement that are very lightweight, I think it does not add some complexity. So the low usability added on our sole interest comes with a low price, it stays balanced. 

What do you think of that?

Reply:
A big part of the reason I'm not a huge fan of this change is having the two functions as return values feels a bit awkward. Are there any alternatives you like better such as a context manager or a simple class with a `workspace` attribute and start and stop methods?

Reply:
Context manager is, I think, my preferred feature in Python, for sure I am willing to implement that.

Done ^^